File: frontend/types/analysis.ts
/**
 * TrueSight — TypeScript Types: Analysis
 * Core types for media analysis results, signals, and scoring.
 *
 * IMPORTANT: The authenticity score is an AI-assisted estimate.
 * It is NOT a scientific certainty and should never be presented as such.
 */

import type { MediaType } from "./media";

// ── Enumerations ──────────────────────────────────────────────────────────────

export type AnalysisStatus = "PENDING" | "PROCESSING" | "COMPLETED" | "FAILED";

/**
 * Overall verdict on the media's authenticity.
 * These should always be presented with appropriate caveats.
 */
export type Verdict =
  | "LIKELY_AUTHENTIC"
  | "SUSPICIOUS"
  | "LIKELY_MANIPULATED"
  | "INCONCLUSIVE";

export type SignalStatus = "positive" | "negative" | "neutral";
export type SignalSeverity = "low" | "medium" | "high";
export type SignalCategory =
  | "visual"
  | "audio"
  | "metadata"
  | "temporal"
  | "ai_pattern"
  | "provenance";

// ── Signal ────────────────────────────────────────────────────────────────────

/**
 * A single detection signal contributing to the overall assessment.
 * Each signal represents one observable characteristic of the media.
 */
export interface AnalysisSignal {
  name: string;
  category: SignalCategory;
  /** positive = supports authenticity, negative = suggests manipulation */
  status: SignalStatus;
  /** Plain-language explanation shown directly to the user. */
  explanation: string;
  severity: SignalSeverity;
}

// ── Analysis Result ───────────────────────────────────────────────────────────

/**
 * The complete result of a single analysis.
 *
 * authenticityScore: 0 (highly suspicious) → 100 (highly authentic).
 * This is an AI-assisted estimate, NOT a scientific fact.
 */
export interface Analysis {
  id: string;
  userId: string | null;
  mediaType: MediaType;
  filename: string | null;
  fileUrl: string | null;
  status: AnalysisStatus;

  /**
   * AI-assisted authenticity estimate (0–100).
   * Lower values indicate higher suspicion of manipulation.
   * Present with appropriate uncertainty language.
   */
  authenticityScore: number | null;
  verdict: Verdict | null;
  /** Model confidence in its assessment (0–1). */
  confidence: number | null;
  /** Plain-language summary for the user. */
  summary: string | null;
  signals: AnalysisSignal[] | null;
  /** Raw technical metadata extracted from the file. */
  metadata: Record<string, unknown> | null;

  createdAt: string;
  updatedAt: string | null;
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/** Human-readable verdict labels. */
export const VERDICT_LABELS: Record<Verdict, string> = {
  LIKELY_AUTHENTIC: "Likely Authentic",
  SUSPICIOUS: "Suspicious",
  LIKELY_MANIPULATED: "Likely Manipulated",
  INCONCLUSIVE: "Inconclusive",
};

/** Returns a color semantic for a given verdict (for UI use). */
export function getVerdictColor(verdict: Verdict): "green" | "yellow" | "red" | "gray" {
  switch (verdict) {
    case "LIKELY_AUTHENTIC":
      return "green";
    case "SUSPICIOUS":
      return "yellow";
    case "LIKELY_MANIPULATED":
      return "red";
    case "INCONCLUSIVE":
      return "gray";
  }
}

/** Converts an authenticity score to a human-readable tier label. */
export function getScoreTier(score: number): string {
  if (score >= 80) return "High Authenticity";
  if (score >= 60) return "Moderate Authenticity";
  if (score >= 40) return "Low Authenticity";
  return "Highly Suspicious";
}

File: backend/app/schemas/analysis.py
"""
TrueSight Backend — Pydantic Schemas: Analysis
Request and response models for analysis endpoints.
Full implementation in STEP 5.
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ── Enumerations ──────────────────────────────────────────────────────────────

class MediaType(str, Enum):
    IMAGE = "IMAGE"
    AUDIO = "AUDIO"
    VIDEO = "VIDEO"
    URL = "URL"


class AnalysisStatus(str, Enum):
    PENDING = "PENDING"
    PROCESSING = "PROCESSING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class Verdict(str, Enum):
    LIKELY_AUTHENTIC = "LIKELY_AUTHENTIC"
    SUSPICIOUS = "SUSPICIOUS"
    LIKELY_MANIPULATED = "LIKELY_MANIPULATED"
    INCONCLUSIVE = "INCONCLUSIVE"


class SignalStatus(str, Enum):
    POSITIVE = "positive"    # Supports authenticity
    NEGATIVE = "negative"    # Suggests manipulation
    NEUTRAL = "neutral"      # No strong signal


class SignalSeverity(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class SignalCategory(str, Enum):
    VISUAL = "visual"
    AUDIO = "audio"
    METADATA = "metadata"
    TEMPORAL = "temporal"
    AI_PATTERN = "ai_pattern"
    PROVENANCE = "provenance"


# ── Signal ────────────────────────────────────────────────────────────────────

class AnalysisSignal(BaseModel):
    """
    A single detection signal contributing to the overall assessment.
    Each signal represents one observable characteristic of the media.
    """

    name: str = Field(..., description="Human-readable signal name")
    category: SignalCategory
    status: SignalStatus
    explanation: str = Field(..., description="Plain-language explanation for the user")
    severity: SignalSeverity = SignalSeverity.LOW


# ── Request Schemas ───────────────────────────────────────────────────────────

class AnalyzeUrlRequest(BaseModel):
    """Request body for URL-based analysis."""

    url: str = Field(..., description="The URL of the media to analyze")
    media_type: Optional[MediaType] = Field(
        None,
        description="Optional hint about the media type; will be inferred if not provided",
    )


# ── Response Schemas ──────────────────────────────────────────────────────────

class AnalysisResult(BaseModel):
    """
    The complete result of a single analysis.
    Authenticity score is AI-assisted and should NOT be treated as definitive truth.
    """

    id: UUID
    user_id: Optional[UUID] = None
    media_type: MediaType
    filename: Optional[str] = None
    file_url: Optional[str] = None
    status: AnalysisStatus

    # Score: 0 (highly suspicious) → 100 (highly authentic).
    # This is an AI-assisted estimate, not a scientific certainty.
    authenticity_score: Optional[float] = Field(
        None, ge=0, le=100, description="AI-assisted authenticity estimate (0–100)"
    )
    verdict: Optional[Verdict] = None
    confidence: Optional[float] = Field(None, ge=0, le=1, description="Model confidence (0–1)")
    summary: Optional[str] = Field(None, description="Plain-language summary for the user")
    signals: Optional[List[AnalysisSignal]] = None
    metadata: Optional[Dict[str, Any]] = Field(
        None, description="Raw technical metadata extracted from the file"
    )
    created_at: datetime
    updated_at: Optional[datetime] = None


class AnalysisListResponse(BaseModel):
    """Paginated list of analyses."""

    items: List[AnalysisResult]
    total: int
    page: int = 1
    page_size: int = 20
